fileseeker.go

```diff
@@ -42,16 +48,147 @@ type statikFile struct {
 	Time    time.Time `json:"time"`
 }
 
+type Config struct {
+	UpdateFrequency int    `toml:"update_frequency"`
+	Port            int    `toml:"port"`
+	DataPath        string `toml:"data_path"`
+	BaseUrl         string `toml:"root_url"`
+}
+
+type LockedFs struct {
+	fs webdav.FileSystem
+	mu *sync.RWMutex
+}
+
+func (lfs *LockedFs) OpenFile(ctx context.Context, name string, flag int, perm os.FileMode) (webdav.File, error) {
+	lfs.mu.RLock()
+
+	file, err := lfs.fs.OpenFile(ctx, name, flag, perm)
+	if err != nil {
+		lfs.mu.RUnlock()
+		return nil, err
+	}
+	return &LockedFile{file: file, mu: lfs.mu}, nil
+}
+
+func (lfs *LockedFs) Mkdir(ctx context.Context, name string, perm os.FileMode) error {
+	return errors.New("read-only filesystem: Mkdir is not allowed")
+}
+
+func (lfs *LockedFs) Rename(ctx context.Context, oldName, newName string) error {
+	return errors.New("read-only filesystem: Rename is not allowed")
+}
+
+func (lfs *LockedFs) RemoveAll(ctx context.Context, name string) error {
+	return errors.New("read-only filesystem: RemoveAll is not allowed")
+}
+
+func (lfs *LockedFs) Stat(ctx context.Context, name string) (os.FileInfo, error) {
+	return lfs.fs.Stat(ctx, name)
+}
+
+// needed to release RWMutex ()
+type LockedFile struct {
+	file webdav.File
+	mu   *sync.RWMutex
+}
+
+func (lf *LockedFile) Close() error {
+	err := lf.file.Close()
+	lf.mu.RUnlock()
+	return err
+}
+
+func (lf *LockedFile) Read(p []byte) (n int, err error) { return lf.file.Read(p) }
+
+func (lf *LockedFile) Write(p []byte) (n int, err error) {
+	return 0, errors.New("read-only filesystem: Write is not allowed")
+}
+
+func (lf *LockedFile) Seek(offset int64, whence int) (int64, error) {
+	return lf.file.Seek(offset, whence)
+}
+
+func (lf *LockedFile) Stat() (fs.FileInfo, error) { return lf.file.Stat() }
+
+func (lf *LockedFile) Readdir(count int) ([]fs.FileInfo, error) { return lf.file.Readdir(count) }
+
 var (
 	dataDir    = flag.String("d", "data", "data directory")
 	configFile = flag.String("c", "config/teachings.json", "config file")
 )
 
+/* middleware to allow only read-only requests */
+func readonlyMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		switch r.Method {
+		case "GET", "PROPFIND", "HEAD", "OPTIONS":
+			next.ServeHTTP(w, r)
+		default:
+			http.Error(w, "405: Method not allowed", http.StatusMethodNotAllowed)
+		}
+	})
+}
+
 func main() {
 	flag.Parse()
 
 	log.Info("Starting fileseeker...")
 
+	var config, err = loadConfig()
+	if err != nil {
+		os.Exit(1)
+	}
+
+	var fsLock sync.RWMutex
+	/* file system */
+	lfs := &LockedFs{fs: webdav.Dir(config.DataPath), mu: &fsLock}
+
+	ticker := time.NewTicker(time.Duration(config.UpdateFrequency) * time.Minute)
+
+	done := make(chan bool)
+
+	/* run statikBFS */
+	go func() {
+		for {
+			select {
+			case <-done:
+				return
+			case <-ticker.C:
+				fsLock.Lock()
+				statikBFS()
+				fsLock.Unlock()
+			}
+		}
+	}()
```

Comment:
Se è la primissima volta che il programma viene lanciato e la cartella `data` non è presente, bisogna aspettare l'intero tempo dell'update prima che vengano scaricati i dati. Bisogna fare in modo che appena parta il programma i dati vengano scaricati subito e poi si inizia a contare l'update frequency.